### PR TITLE
Reduce likelihood of DRS hogging IoActor NIO threads [BT-341]

### PR DESCRIPTION
**Pull request by mcovarr**
The `DrsCloudNioFileSystemProvider` was wrapping the retries of the `DrsPathResolver` with another set of `CloudNioRetry` retries. The product of these two retries at the previous configuration values would wait around 35 minutes (~:20 + 10 x ~3:30) to fail for each doomed attempt. That combined with a fairly wide scatter and a typo'd DRS path for `file` in code like 

```
task size {
    input {
        File file
    }

    Int file_size = ceil(size(file))
...
}
```
would completely block all 10 of the `IoActor`s [NIO threads](https://github.com/broadinstitute/cromwell/blob/develop/engine/src/main/scala/cromwell/server/CromwellRootActor.scala#L108).

These changes remove the nested retries in the engine and dial back the patience for retries. If we want the retries to be more patient we'll probably have to make other code that is competing for `IoActor` threads more patient as well.

Utility files for reproducing this error can be cherry picked from commit `ff7bddc8830802f7a606177d0eaf19c8f47ca865`. I don't know how to programatically link Google accounts to NIH accounts in Bond to be able to include this Centaur test in CI, though maybe we don't need to be linked to make sure this negative case errors within a reasonable timeout?

Workflow`9635fbf0-00b1-4635-b482-5a782cda5cd5` induced this problem in production, its metadata shows multiple `HaplotypeCaller` shards erroring out with 
```
Failed to evaluate input 'disk_size' (reason 1 of 1): [Attempted 1 time(s)] - RuntimeException: Unexpected response during DRS resolution: RuntimeException: Could not access object 'drs://dg4.DFC/...'. Status: 500, reason: 'Internal Server Error', Martha location: 'https://.../martha_v3', message: 'Received error while resolving DRS URL. getaddrinfo ENOTFOUND dg4.dfc'
```

### CODE REVIEW

File: cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala [hunk 1]
```diff
@@ -28,7 +29,7 @@ class CloudNioRetry(config: Config) {
       case Failure(exception: Exception) if isFatal(exception) => throw exception
       case Failure(exception: Exception) if !isFatal(exception) =>
         val retriesLeft = if (isTransient(exception)) maxRetries else maxRetries map { _ - 1 }
-        if (retriesLeft.forall(_ > 0)) {
+        if (retriesLeft.forall(_ >= 0)) {
```
Comment (mcovarr) [inline cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala:32]: fixes and off by 1 issue

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsConfig.scala [hunk 2]
```diff
@@ -16,9 +17,9 @@ final case class DrsConfig(marthaUrl: String,
 object DrsConfig {
   // If you update these values also update Filesystems.md!
   private val DefaultNumRetries = 3
-  private val DefaultWaitInitial = 30.seconds
-  private val DefaultWaitMaximum = 5.minutes
-  private val DefaultWaitMultiplier = 2.0d
+  private val DefaultWaitInitial = 10 seconds
+  private val DefaultWaitMaximum = 30 seconds
+  private val DefaultWaitMultiplier = 1.5d
   private val DefaultWaitRandomizationFactor = 0.1
```
Comment (mcovarr) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsConfig.scala:23]: These values are actually not relevant in this PR since the inner retry is not invoked through the engine code; I dialed them back anyway b/c they seemed quite long.

File: cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala [hunk 3]
```diff
@@ -4,17 +4,18 @@ import com.typesafe.config.Config
...
 
-  def defaultMaxRetries: Int = config.getAs[Int]("default-max-retries").getOrElse(10)
+  def defaultMaxRetries: Int = config.getAs[Int]("default-max-retries").getOrElse(3)
 
   def isFatal(exception: Exception): Boolean = !isTransient(exception)
 
   def isTransient(exception: Exception): Boolean = false
 
-  def defaultBackOff: CloudNioBackoff = CloudNioSimpleExponentialBackoff(1.second, 60.seconds, 1.1)
+  def defaultBackOff: CloudNioBackoff = CloudNioSimpleExponentialBackoff(2 seconds, 8 seconds, 2)
```
Comment (mcovarr) [inline cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala:18]: The total wait time is actually not too different between these two settings (~18 s vs ~14s), though the number of potential calls to Martha is much less.

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsConfig.scala (hunk 2 above)
Comment (kshakir) [reply cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsConfig.scala:23]: They were long because due to WA-90 Martha v2 & v3 already re-hammers servers when they report that they are overloaded. Shortening the time here shrinks the amount of time given to the servers to recover, and possibly finishing a restart if they've crashed. For example, I think when we knocked over JDR in the past it took around 2 minutes for the application to restart.

That said, I don't have any current empirical evidence to support NOT changing these numbers, so LGTM

File: cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala [hunk 4]
```diff
@@ -4,17 +4,18 @@ import com.typesafe.config.Config
 import net.ceedubs.ficus.Ficus._
 
 import scala.concurrent.duration._
+import scala.language.postfixOps
 import scala.util.{Failure, Success, Try}
 
 class CloudNioRetry(config: Config) {
 
-  def defaultMaxRetries: Int = config.getAs[Int]("default-max-retries").getOrElse(10)
+  def defaultMaxRetries: Int = config.getAs[Int]("default-max-retries").getOrElse(3)
```
Comment (kshakir) [inline cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala:12]: Again, no strong empirical evidence, but I believe this "10" was copied from the behavior of the built-in Papi V1 localizer / delocalizer that retried that number of times.

This does appear to be touching an SPI (thus a common interface) and not an implementation, but still LGTM

File: docs/filesystems/Filesystems.md [hunk 5]
```diff
@@ -22,10 +22,10 @@ filesystems {
...
             # The number of times to retry failures connecting or HTTP 429 or HTTP 5XX responses, default 3.
             num-retries = 3
-            # How long to wait between retrying HTTP 429 or HTTP 5XX responses, default 30 seconds.
-            wait-initial = 30 seconds
-            # The maximum amount of time to wait between retrying HTTP 429 or HTTP 5XX responses, default 5 minutes.
-            wait-maximum = 5 minutes
+            # How long to wait between retrying HTTP 429 or HTTP 5XX responses, default 10 seconds.
+            wait-initial = 10 seconds
+            # The maximum amount of time to wait between retrying HTTP 429 or HTTP 5XX responses, default 30 seconds.
+            wait-maximum = 30 seconds
```
Comment (kshakir) [inline docs/filesystems/Filesystems.md:28]: Again, 30 seconds is short for an overloaded server to crash and start back up. And I'm not convinced a lot of DRS implementations are load balanced / high availablity.

**Review by kshakir — APPROVED** [review]
These changes seem ancillary to the as-defined BT-313. Still I'm giving this a LGTM for the needs of the many users NOT using this code path that may be affected by ripples emanating from this area and producing repeated BT-313 issues.

File: cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala (hunk 4 above)
Comment (mcovarr) [reply cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala:12]: oh excellent point, I think I may revert this bit then

Comment (breilly2) [reply cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala:12]: Just agreeing that we shouldn't change this if it affects things more universally, especially if we're not sure if the potential benefit is worth the potential risk of collateral effects.

File: cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala (hunk 1 above)
Comment (breilly2) [reply cloud-nio/cloud-nio-spi/src/main/scala/cloud/nio/spi/CloudNioRetry.scala:32]: Assuming the original intent was "max retries" rather than "max tries". :)

Good catch!

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsConfig.scala (hunk 2 above)
Comment (breilly2) [reply cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsConfig.scala:23]: I'm sure there was good reason for these values in the past, and that reason may not have changed. The thing that tips me over to agreeing to this change is that _not_ making this change could put Cromwell (or at least other workflows) in jeopardy, which feels like the wrong trade-off for being _very_ patient with DRS providers falling over and having to restart. Shortening this helps protect Cromwell while putting the pressure on other services to behave more reasonably.

Ideally, we could be patient _and_ not jeopardize overall Cromwell health, but that's an effort for another day.

**Review by breilly2 — APPROVED** [review]
Agreeing with the consensus to rollback the change to `CloudNioRetry.defaultMaxRetries`. Otherwise, 👍
